src/sentry/models/releasefile.py

```diff
@@ -177,45 +186,165 @@ def extract(self) -> TemporaryDirectory:
         return temp_dir
 
 
-def merge_release_archives(file1: IO, archive2: ReleaseArchive, target: IO) -> bool:
-    """Append contents of archive2 to copy of file1
+class _ArtifactIndexData:
+    """Holds data of artifact index and keeps track of changes"""
+
+    def __init__(self, data: dict, fresh=False):
+        self._data = data
+        self.changed = fresh
+
+    @property
+    def data(self):
+        """Meant to be read-only"""
+        return self._data
+
+    def get(self, filename: str):
+        return self._data.get("files", {}).get(filename, None)
+
+    def update_files(self, files: dict):
+        if files:
+            self._data.setdefault("files", {}).update(files)
+            self.changed = True
+
+    def delete(self, filename: str):
+        self._data.get("files", {}).pop(filename, None)
+        self.changed = True
+
+
+class _ArtifactIndexGuard:
+    """Ensures atomic write operations to the artifact index"""
+
+    def __init__(self, release: Release, dist: Optional[Distribution]):
+        self._release = release
+        self._dist = dist
+
+    def readable_data(self) -> Optional[dict]:
+        """Simple read, no synchronization necessary"""
+        file_ = self._get_file(lock=False)
+        if file_ is not None:
+            with file_.getfile() as fp:
+                return json.load(fp)
+
+    @contextmanager
+    def writable_data(self, create: bool) -> ContextManager[_ArtifactIndexData]:
+        """Context manager for editable artifact index"""
+        with transaction.atomic():
+            if create:
+                file_, created = self._get_or_create_file()
+            else:
+                file_ = self._get_file(lock=True)
+                created = False
+
+            if file_ is None:
+                index_data = None
+            else:
+                if created:
+                    index_data = _ArtifactIndexData({}, fresh=True)
+                else:
+                    raw_data = json.load(file_.getfile())
+                    index_data = _ArtifactIndexData(raw_data)
+
+            yield index_data  # editable reference to index
+
+            if index_data is not None and index_data.changed:
+                file_.putfile(BytesIO(json.dumps(index_data.data).encode()))
+
+    def _get_or_create_file(self) -> Tuple[File, bool]:
+        # Make sure the appropriate rows are locked for update:
+        qs = ReleaseFile.objects.select_related("file").select_for_update()
+
+        file_ = File.objects.create(
```

Comment:
Can you add a code comment pointing out that the file is to be populated by the caller?  And probably also best to call out in the caller what effect the `select_for_update` has.